algebird-core/src/main/scala/com/twitter/algebird/TDigest.scala

```diff
@@ -0,0 +1,263 @@
+      val clustNew = cmNew.foldLeft(clustDel)((c, p) => c.increment(p._1, p._2))
+      val nc = nclusters - s.length + cmNew.length
+
+      // return the updated t-digest
+      TDigest(delta, nc, clustNew)
+    }
+  }
+
+  /**
+   * Compute a cumulative probability (CDF) for a numeric value, from the estimated probability
+   * distribution represented by this t-digest sketch.
+   * @param x a numeric value
+   * @return the cumulative probability that a random sample from the distribution is <= x
+   */
+  def cdf[N](x: N)(implicit num: Numeric[N]): Double = clusters.cdf(x)
+
+  /**
+   * Compute the inverse cumulative probability (inverse-CDF) for a quantile value, from the
+   * estimated probability distribution represented by this t-digest sketch.
+   * @param q a quantile value.  The value of q is expected to be on interval [0, 1]
+   * @return the value x such that cdf(x) = q
+   */
+  def cdfInverse[N](q: N)(implicit num: Numeric[N]): Double = clusters.cdfInverse(q)
+}
+
+/** Factory functions for TDigest */
+object TDigest {
+  /**
+   * Default value for a t-digest delta parameter.  The number of clusters varies, roughly, as
+   * about (50/delta), when data are presented in random order
+   * (it may grow larger if data are not presented randomly).  The default corresponds to
+   * an expected number of clusters of about 100.
+   */
+  val deltaDefault: Double = (50.0 / 100.0) // delta * E[clusters] ~ 50
+
+  /**
+   * The t-digest algorithm will re-cluster itself whenever its number of clusters exceeds
+   * (K/delta).  This value is set such that the threshold is about 10x the heuristically
+   * expected number of clusters for the user-specified delta value.  Generally the number of
+   * clusters will only trigger the corresponding re-clustering threshold when data are being
+   * presented in a non-random order.
+   */
+  val K: Double = 10.0 * 50.0
+
+  /**
+   * Obtain an empty t-digest
+   * @param delta a sketch resolution parameter.
+   * @note Smaller values of delta yield sketches with more clusters, and higher resolution
+   * @note The expected number of clusters will vary (roughly) as (50/delta)
+   */
+  def empty(delta: Double = deltaDefault): TDigest = {
+    require(delta > 0.0, s"delta was not > 0")
+    TDigest(delta, 0, TDigestMap.empty)
+  }
+
+  /**
+   * Sketch some data with a t-digest
+   * @param data The data elements to sketch
+   * @param delta The sketch resolution parameter.
+   * @return A t-digest sketch of the input data
+   * @note Smaller values of delta yield sketches with more clusters, and higher resolution
+   * @note The expected number of clusters will vary (roughly) as (50/delta)
+   */
+  def sketch[N](
+    data: TraversableOnce[N],
+    delta: Double = deltaDefault)(implicit num: Numeric[N]): TDigest = {
+    require(delta > 0.0, s"delta was not > 0")
+    val td = data.foldLeft(empty(delta))((c, e) => c + ((e, 1)))
+    val ds = scala.util.Random.shuffle(td.clusters.toVector)
```

Comment:
here is another randomization. Is this needed? Why not have the user shuffle if they want that behavior and just do the fold here?